doc:  update README (#84)
Azure/notation-azure-kv: README.md

[![codecov](https://codecov.io/gh/Azure/notation-azure-kv/branch/main/graph/badge.svg)](https://codecov.io/gh/Azure/notation-azure-kv)

Azure Provider for the [Notation CLI](https://github.com/notaryproject/notation)

> The notation-azure-kv plugin provides the capability to sign the Notation generated payload by using Azure Key Vault (AKV). The user's certificate and private key should be stored in AKV and the plugin will request signing and getting the leaf certificate from AKV. 

Reviewer: You probably want to add back ticks to all occurrences of `notation-azure-kv`

> The notation-azure-kv plugin provides the capability to sign the Notation generated payload by using Azure Key Vault (AKV). The user's certificate and private key should be stored in AKV and the plugin will request signing and getting the leaf certificate from AKV. 

Author: Updated.

> The plugin supports Azure CLI identity and Managed Identity for accessing AKV.

Reviewer: Are you referring to [AAD workload identities](https://learn.microsoft.com/azure/active-directory/develop/workload-identities-overview) for "Azure CLI identity and Managed Identity"?

Author: Azure CLI identity and MI token are the 2 supported itentities for the current version. workload identity is not supported.

Reviewer: Can you provide a link? otherwise, customers will be confused.

Author: Updated.

## Installation
Before installing notation azure key vault plugin, please make sure Notation CLI has been installed. If you didn't install it, please follow the [Notation installation guide](https://notaryproject.dev/docs/installation/cli/).

1. Navigate to the [Releases](https://github.com/Azure/notation-azure-kv/releases) page and select the latest release of notation-azure-kv. The notation-azure-kv binaries for each platform are available under the Assets section. Please download it based on your own platform. 
2. Validate the checksum which should be the same as the one in `checksums.txt` and then install the plugin.
   
   For Linux Bash:
   ```sh

>    # validate checksum
>    sha256sum notation-azure-kv_<version>_linux_amd64.tar.gz

Reviewer: It is suggested to do
```suggestion
   version=v0.6.0

   # validate checksum
   cat checksums.txt | grep notation-azure-kv_${version}_linux_amd64.tar.gz | sha256sum -c
```

Reviewer: Again, use variables instead of placeholders.

Author: Updated.

   # install the plugin
   mkdir -p "$HOME/.config/notation/plugins/azure-kv"
   tar zxf notation-azure-kv_<version>_linux_amd64.tar.gz -C "$HOME/.config/notation/plugins/azure-kv" notation-azure-kv
   ```

>    For MacOS Bash:

Reviewer: Mac use Zsh.
```suggestion
   For macOS Zsh:
```

Author: Updated.

   ```sh
   # validate checksum
   shasum -a 256 notation-azure-kv_<version>_linux_amd64.tar.gz

   # install the plugin
   mkdir -p "$HOME/Library/Application Support/notation/plugins/azure-kv"
   tar zxf notation-azure-kv_<version>_darwin_amd64.tar.gz -C "$HOME/Library/Application Support/notation/plugins/azure-kv" notation-azure-kv
   ```
   For Windows Powershell:
   ```powershell
   # validate checksum
   (Get-FileHash .\notation-azure-kv_<version>_windows_amd64.zip).Hash

   # install the plugin
   mkdir "$env:AppData\notation\plugins\azure-kv"
   Expand-Archive -Path notation-azure-kv_<version>_windows_amd64.zip -DestinationPath "$env:AppData\notation\plugins\azure-kv"
   ```
3. Try to run `notation plugin list` to show the installed plugin.

## Getting started with a self-signed Azure Key Vault Certificate

> To demonstrate the basic use case, starting with a self-signed certificate is easier to understand how the Azure-kv plugin works with Notation, however, you **should not** use this in production because the self-signed certificate is not trusted.

Reviewer: You may want to remove this line since self-signed certificates can be trusted depending on the scenarios.

Of course, you may replace it with a note:

> **Note** It is suggested to get a certificate from a trusted CA since self-signed certificate is not publicly trusted.

   keyVault=notationKV
   location=westus
   certName=notationSelfSignedCert

   # create a resource group
   az group create -n $resourceGroup -l $location
   
   # create a Azure Key Vault
   az keyvault create -l $location -n $keyVault --resource-group $resourceGroup

   # generate certificate policy
   cat <<EOF > ./selfSignedPolicy.json
   {
     "issuerParameters": {
       "certificateTransparency": null,
       "name": "Self"
     },
     "keyProperties": {
       "curve": null,
       "exportable": false,
       "keySize": 2048,
       "keyType": "RSA",
       "reuseKey": true
     },
     "secretProperties": {
       "contentType": "application/x-pem-file"
     },
     "x509CertificateProperties": {
        "ekus": [
           "1.3.6.1.5.5.7.3.3"
       ],
       "keyUsage": [
         "digitalSignature"
       ],
       "subject": "CN=Test Signer",
       "validityInMonths": 12
     }
   }
   EOF

   # create self-signed certificate
   az keyvault certificate create -n $certName --vault-name $keyVault -p @selfSignedPolicy.json

   # get the key identifier
   keyID=$(az keyvault certificate show -n $certName --vault-name $keyVault --query 'kid' -o tsv)
   ```

> 4. Prepare a container registry. You can follow the [guide](https://learn.microsoft.com/en-us/azure/container-registry/container-registry-get-started-portal?tabs=azure-cli) to create an Azure Container Registry which is recommended. 

Reviewer: nit: make link not language specific
```suggestion
4. Prepare a container registry. You can follow the [guide](https://learn.microsoft.com/azure/container-registry/container-registry-get-started-portal?tabs=azure-cli) to create an Azure Container Registry which is recommended. 
```

> 5. Sign the container image with Notation:
>    ```sh
>    # sign with azure-kv
>    notation sign localhost:5000/hello-world:v1 --id $keyID --plugin azure-kv
>    # example output
>    Warning: Always sign the artifact using digest(@sha256:...) rather than a tag(:v1) because tags are mutable and a tag reference can point to a different artifact than the one signed.
>    Successfully signed localhost:5000/hello-world@sha256:f54a58bc1aac5ea1a25d796ae155dc228b3f0e11d046ae276b39c4bf2f13d8c4
>    ```

Reviewer: ````suggestion
5. Sign the container image using `notation` and `notation-azure-kv`:
   ```sh
   notation sign localhost:5000/hello-world:v1 --id $keyID --plugin azure-kv
   ```
   Example output:
   ```
   Warning: Always sign the artifact using digest(@sha256:...) rather than a tag(:v1) because tags are mutable and a tag reference can point to a different artifact than the one signed.
   Successfully signed localhost:5000/hello-world@sha256:f54a58bc1aac5ea1a25d796ae155dc228b3f0e11d046ae276b39c4bf2f13d8c4
   ```
````

Author: Updated.

6. Verify the signature associated with the image:
   ```sh
   # add self-signed certificate to notation trust store
   cat <<EOF > ./selfSignedCert.crt
   -----BEGIN CERTIFICATE-----
   $(az keyvault certificate show -n $certName --vault-name $keyVault --query 'cer' -o tsv)
   -----END CERTIFICATE-----
   EOF
   notation cert add --type ca --store selfSigned ./selfSignedCert.crt
   
   # add notation trust policy
   notationConfigDir="${HOME}/.config/notation"                      # for Linux

>    notationConfigDir="${HOME}/Library/Application Support/notation"  # for macOS

Reviewer: Comment out the macOS one so that the block of code can be copied.
```suggestion
   #notationConfigDir="${HOME}/Library/Application Support/notation"  # for macOS
```

   mkdir -p $notationConfigDir
   cat <<EOF > $notationConfigDir/trustpolicy.json
   {
    "version": "1.0",
    "trustPolicies": [
        {
            "name": "trust-policy-example",
            "registryScopes": [ "*" ],
            "signatureVerification": {
                "level" : "strict" 
            },
            "trustStores": [ "ca:selfSigned" ],
            "trustedIdentities": [
                "*"
            ]
        }
    ]
   }
   EOF

>    chmod 0600 $notationConfigDir/trustpolicy.json

Reviewer: nit:
```suggestion
   chmod 600 $notationConfigDir/trustpolicy.json
```

Author: Updated.

   # verify signature
   notation verify localhost:5000/hello-world:v1
   # example output
   Warning: Always verify the artifact using digest(@sha256:...) rather than a tag(:v1) because resolved digest may not point to the same signed artifact, as tags are mutable.

>    Successfully verified signature for localhost:5000/hello-world@sha256:f54a58bc1aac5ea1a25d796ae155dc228b3f0e11d046ae276b39c4bf2f13d8c4

Reviewer: Move out the example output as above

   ```

> ## Getting started with a certificate signed by a trusted CA.

Reviewer: nit:
```suggestion
## Getting started with a certificate signed by a trusted CA
```

Author: Updated.

   
   # create a Azure Key Vault
   az keyvault create -l $location -n $keyVault --resource-group $resourceGroup

   # generate certificate policy
   cat <<EOF > ./leafCert.json
   {
     "issuerParameters": {
       "certificateTransparency": null,
       "name": "Unknown"
     },
     "keyProperties": {
       "curve": null,
       "exportable": false,
       "keySize": 2048,
       "keyType": "RSA",
       "reuseKey": true
     },
     "secretProperties": {
       "contentType": "application/x-pem-file"
     },
     "x509CertificateProperties": {
       "ekus": [
           "1.3.6.1.5.5.7.3.3"
       ],
       "keyUsage": [
         "digitalSignature"
       ],
       "subject": "CN=Test Signer",
       "validityInMonths": 12
     }
   }
   EOF

   # create the leaf certificate
   az keyvault certificate create -n $certName --vault-name $keyVault -p @leafCert.json

   # get the CSR
   CSR=$(az keyvault certificate pending show --vault-name $keyVault --name $certName --query 'csr' -o tsv)
   CSR_PATH=${certName}.csr
   printf -- "-----BEGIN CERTIFICATE REQUEST-----\n%s\n-----END CERTIFICATE REQUEST-----\n" $CSR > ${CSR_PATH}
   ```

> 4. Please take ${certName}.csr file to a trusted CA to sign and issue your certificate, or you can use `openssl` tool to sign it locally for testing.

Reviewer: ```suggestion
4. Please take `${certName}.csr` file to a trusted CA to sign and issue your certificate, or you can use `openssl` tool to sign it locally for testing.
```

   Successfully signed localhost:5000/hello-world@sha256:f54a58bc1aac5ea1a25d796ae155dc228b3f0e11d046ae276b39c4bf2f13d8c4
   ```
8. Signature verification with Notation needs the root certificate of your CA in the trust store:
   ```sh
   # add root certificate ($rootCertPath) to notation trust store
   notation cert add --type ca --store trusted $rootCertPath
   
   # add notation trust policy
   notationConfigDir="${HOME}/.config/notation"                      # for Linux
   notationConfigDir="${HOME}/Library/Application Support/notation"  # for macOS

   mkdir -p $notationConfigDir
   cat <<EOF > $notationConfigDir/trustpolicy.json
   {
    "version": "1.0",
    "trustPolicies": [
        {
            "name": "trust-policy-example",
            "registryScopes": [ "*" ],
            "signatureVerification": {
                "level" : "strict" 
            },
            "trustStores": [ "ca:trusted" ],
            "trustedIdentities": [
                "*"
            ]
        }
    ]
   }
   EOF

>    chmod 0600 $notationConfigDir/trustpolicy.json

Reviewer: ```suggestion
   chmod 600 $notationConfigDir/trustpolicy.json
```

Author: Updated.

   # verify signature
   notation verify localhost:5000/hello-world:v1
   # example output
   Warning: Always verify the artifact using digest(@sha256:...) rather than a tag(:v1) because resolved digest may not point to the same signed artifact, as tags are mutable.
   Successfully verified signature for localhost:5000/hello-world@sha256:f54a58bc1aac5ea1a25d796ae155dc228b3f0e11d046ae276b39c4bf2f13d8c4

>    `

Reviewer: nit: typo
````suggestion
   ```
````

> > Please make sure the certificate is in PEM format. PCKS#12 will be supported in the future.

Reviewer: ```suggestion
> **Note** Please make sure the certificate is in PEM format. PCKS#12 will be supported in the future.
```

Reviewer: `**Note**` becomes
> **Note**

`**Warning**` becomes
> **Warning**

Author: > **Note**

> **Warning** 

Cool.

[![codecov](https://codecov.io/gh/Azure/notation-azure-kv/branch/main/graph/badge.svg)](https://codecov.io/gh/Azure/notation-azure-kv)

Azure Provider for the [Notation CLI](https://github.com/notaryproject/notation)

> The `notation-azure-kv` plugin provides the capability to sign the Notation generated payload by using Azure Key Vault (AKV). The user's certificate and private key should be stored in AKV and the plugin will request signing and getting the leaf certificate from AKV. 

Author: Updated.

> The plugin supports authentication by [Azure CLI](https://learn.microsoft.com/cli/azure/authenticate-azure-cli) or [Managed Identity](https://learn.microsoft.com/azure/active-directory/managed-identities-azure-resources/overview). To enable `Managed Identity` authentication, the `AKV_AUTH_METHOD=AKV_AUTH_FROM_MI` should be set in the environment variable.

Author: Updated.

> ## Installation

Author: Updated.

> Before installing notation azure key vault plugin, please make sure Notation CLI has been installed. If you didn't install it, please follow the [Notation installation guide](https://notaryproject.dev/docs/installation/cli/).

Author: Updated.

> 1. Navigate to the [Releases](https://github.com/Azure/notation-azure-kv/releases) page and select the latest release of `notation-azure-kv`. The `notation-azure-kv` binaries for each platform are available under the Assets section. Please download it based on your own platform. 

Author: Updated.

> 2. Validate the checksum which should be the same as the one in `checksums.txt` and then install the plugin.

Author: Updated.

   
   For Linux Bash:
   ```sh
   version=0.6.0

   # validate checksum
   cat checksums.txt | grep notation-azure-kv_${version}_linux_amd64.tar.gz | sha256sum -c

   # install the plugin
   mkdir -p "$HOME/.config/notation/plugins/azure-kv"
   tar zxf notation-azure-kv_<version>_linux_amd64.tar.gz -C "$HOME/.config/notation/plugins/azure-kv" notation-azure-kv
   ```
   For macOS Zsh:
   ```sh
   version=0.6.0

   # validate checksum
   cat checksums.txt | grep notation-azure-kv_${version}_darwin_amd64.tar.gz | shasum -a 256 -c

   # install the plugin
   mkdir -p "$HOME/Library/Application Support/notation/plugins/azure-kv"

>    tar zxf notation-azure-kv_<version>_darwin_amd64.tar.gz -C "$HOME/Library/Application Support/notation/plugins/azure-kv" notation-azure-kv

Author: Updated.

   ```
   For Windows Powershell:
   ```powershell

>    $version = 0.6.0

Author: Updated.

   Expand-Archive -Path notation-azure-kv_<version>_windows_amd64.zip -DestinationPath "$env:AppData\notation\plugins\azure-kv"
   ```
3. Try to run `notation plugin list` to show the installed plugin.

> ## Getting Started with a self-signed Azure Key Vault Certificate

Reviewer 2: ```suggestion
## Getting started with a self-signed certificate in AKV
```

Author: Updated.

>    (Get-FileHash .\notation-azure-kv_$version_windows_amd64.zip).Hash

Author: Updated.

To demonstrate the basic use case, starting with a self-signed certificate is easier to understand how the Azure-kv plugin works with Notation, however, you **should not** use this in production because the self-signed certificate is not trusted.

> 1. install Azure CLI by following the [guide](https://learn.microsoft.com/en-us/cli/azure/install-azure-cli)

Author: Updated.

   ```
3. Try to run `notation plugin list` to show the installed plugin.

> ## Getting started with a self-signed Azure Key Vault Certificate

Author: Updated.

> 2. login with Azure CLI, set the subscription and make sure the `GetCertificate` and `Sign` permission have been granted to your role:

Author: Updated.

> > **Note** It is suggested to get a certificate from a trusted CA since a self-signed certificate is not publicly trusted.

Author: Updated.

> 1. Install Azure CLI by following the [guide](https://learn.microsoft.com/en-us/cli/azure/install-azure-cli)

Author: Updated.

> 2. Login to Azure with Azure CLI, set the subscription and make sure the `GetCertificate` and `Sign` permission have been granted to your role:

Author: Updated.

   ```sh
   az login
   az account set --subscription $subscriptionID
   ```
3. Create an Azure Key Vault and a self-signed certificate:
   ```sh
   resourceGroup=notationResource

>    keyVault=notationKV

Reviewer 3: "The vault name 'notationKV' is already in use. Vault names are globaly unique so it is possible that the name is already taken." 
Maybe we should suggest users to choose own Vault name

Reviewer 4: Agreed,  add a comment stating this must be a unique value and split the KV creation and cert creation into separate steps. If the user has a copy/paste fail b/c they didn't update the value, it'll be easier to troubleshoot.

Author: Updated.

   location=westus
   certName=notationSelfSignedCert

   # create a resource group
   az group create -n $resourceGroup -l $location
   
   # create a Azure Key Vault
   az keyvault create -l $location -n $keyVault --resource-group $resourceGroup

   # generate certificate policy
   cat <<EOF > ./selfSignedPolicy.json
   {
     "issuerParameters": {
       "certificateTransparency": null,
       "name": "Self"
     },
     "keyProperties": {
       "curve": null,
       "exportable": false,
       "keySize": 2048,
       "keyType": "RSA",
       "reuseKey": true
     },
     "secretProperties": {
       "contentType": "application/x-pem-file"
     },
     "x509CertificateProperties": {
        "ekus": [
           "1.3.6.1.5.5.7.3.3"
       ],
       "keyUsage": [
         "digitalSignature"
       ],

>        "subject": "CN=Test Signer",

Reviewer 3: It is not correct per notation requirement, see error
ERRO[2023-03-15T16:35:40+08:00] authenticity validation failed. Failure reason: error while parsing the certificate subject from the digital signature. error : "distinguished name (DN) \"CN=Test Signer\" has no mandatory RDN attribute for \"C\", it must contain 'C', 'ST', and 'O' RDN attributes at a minimum"

Due to `trustidentities` is set to "*", so the check is skipped. Maybe this is a bug?

Author: Updated to include 'C', 'ST', and 'O'

   az keyvault certificate create -n $certName --vault-name $keyVault -p @selfSignedPolicy.json

   # get the key identifier
   keyID=$(az keyvault certificate show -n $certName --vault-name $keyVault --query 'kid' -o tsv)
   ```
4. run a local registry and push an image to be signed:
   ```sh
   # run a local registry with Docker
   docker run --rm -d -p 5000:5000 ghcr.io/oras-project/registry:v1.0.0-rc.4
   
   # push a hello-world image
   docker pull hello-world
   docker tag hello-world:latest localhost:5000/hello-world:v1
   docker push localhost:5000/hello-world:v1
   ```

> 5. notation sign:

Author: Updated.

> 4. Prepare a container registry. You can follow the [guide](https://learn.microsoft.com/azure/container-registry/container-registry-get-started-portal?tabs=azure-cli) to create an Azure Container Registry which is recommended. Suppose the logging server is `notation.azurecr.io` with an `hello-world:v1` image.

Reviewer 4: ```suggestion
4. [Create an Azure Container Registry](https://learn.microsoft.com/azure/container-registry/container-registry-get-started-portal?tabs=azure-cli) for storing your signed images. The remaining steps use the example login server `notation.azurecr.io`, but you must substitute your own login server value.
```

Reviewer 4: Also, you need to add a step giving them a sample image to sign and how to pull it into ACR. If the user is creating an ACR from scratch, they likely don't have an image their ACR to sign.

Author: Updated.

> 4. Prepare a container registry. You can follow the [guide](https://learn.microsoft.com/azure/container-registry/container-registry-get-started-portal?tabs=azure-cli) to create an Azure Container Registry which is recommended. Suppose the logging server is `notation.azurecr.io` with an `hello-world:v1` image.

Author: It's a typo. I want to say "login server" which is the host name of the registry.

> 5. Sign the container image with Notation:

Reviewer 3: missing the `notation login` or `az acr login` into the registry

>    notation sign notation.azurecr.io/hello-world:v1 --id $keyID --plugin azure-kv

Reviewer 5: We need an example of user passing the cert_bundle from notation to akv-plugin through `--plugin-config`.
In addition, since Notation CLI recommends
```
notation key add ...
notation sign ...
```
as default workflow for signing an artifact, I suggest akv-plugin follow the same procedure in its README so that everything would be aligned.

Author: Updated.

>    example output

Reviewer 4: ```suggestion
 The following example output shows the artifact is successfully signed.
```

Author: Updated.

   notation sign localhost:5000/hello-world:v1 --id $keyID --plugin azure-kv
   # example output
   Warning: Always sign the artifact using digest(@sha256:...) rather than a tag(:v1) because tags are mutable and a tag reference can point to a different artifact than the one signed.
   Successfully signed localhost:5000/hello-world@sha256:f54a58bc1aac5ea1a25d796ae155dc228b3f0e11d046ae276b39c4bf2f13d8c4
   ```

> 6. notation verify:

Reviewer 2: ```suggestion
6. Verify the signature associated with the image:
```

Author: Updated.

> 6. Verify the signature associated with the image:

Reviewer 4: ```suggestion
6. Add the certificate to your trust store and verify the signature associated with the image is valid:
```

Reviewer 4: See comment below about splitting this into two different steps.

Author: Updated.

   # add self-signed certificate to notation trust store
   cat <<EOF > ./selfSignedCert.crt
   -----BEGIN CERTIFICATE-----
   $(az keyvault certificate show -n $certName --vault-name $keyVault --query 'cer' -o tsv)
   -----END CERTIFICATE-----
   EOF
   notation cert add --type ca --store selfSigned ./selfSignedCert.crt
   
   # add notation trust policy

>    notationConfigDir="${HOME}/.config/notation"                      # for Linux
>    # notationConfigDir="${HOME}/Library/Application Support/notation"  # for macOS

Reviewer: nit:
```suggestion
   notationConfigDir="${HOME}/.config/notation"                        # for Linux
   # notationConfigDir="${HOME}/Library/Application Support/notation"  # for macOS
```

Author: Updated.

>    notationConfigDir="${HOME}/Library/Application Support/notation"  # for MacOS

Author: Updated.

> 

Reviewer 2: Why Windows is missing here?

Author: Because the guide are written in bash script which is supported in Linux and MacOS. If I want to also include Windows Powershell script, many lines will be different for Bash and Powershell.

Reviewer 4: There's a bit of a disconnect here. Earlier in the article, you have PowerShell examples, but then go only Bash. If a user has been running through this using PowerShell, they can't easily proceed. 

You should add PowerShell versions for the steps, or confirm it works in WSL and call that out.

Author: Updated to mention that the script supports Windows WSL.

   cat <<EOF > $notationConfigDir/trustpolicy.json
   {
    "version": "1.0",
    "trustPolicies": [
        {
            "name": "trust-policy-example",
            "registryScopes": [ "*" ],
            "signatureVerification": {
                "level" : "strict" 
            },
            "trustStores": [ "ca:selfSigned" ],
            "trustedIdentities": [
                "*"
            ]
        }
    ]
   }
   EOF
   chmod 600 $notationConfigDir/trustpolicy.json
   ```

>    verify signature

Reviewer 4: I think you should split the "add to trust store" and "verify" into separate numbered steps.

Author: Updated

   ```
   notation verify notation.azurecr.io/hello-world:v1
   ```

>    example output

Reviewer 4: ```suggestion
    The following output shows the artifact is successfully verified.
```

Author: Updated.

   ```

> ## Getting Started with a certificate signed by a trusted CA.

Reviewer 2: ```suggestion
## Getting started with a certificate signed by a trusted CA.
```

Author: Updated.

1. install Azure CLI by following the [guide](https://learn.microsoft.com/en-us/cli/azure/install-azure-cli)

> 2. login with Azure CLI, set the subscription and make sure the `GetCertificate` and `Sign` permission have been granted to your role:

Reviewer 2: ```suggestion
2. log in to Azure with Azure CLI, set the subscription, and make sure the `GetCertificate` and `Sign` permission have been granted to your role:
```

Author: Updated.

> ## Getting started with a certificate signed by a trusted CA

Reviewer 4: ```suggestion
## Sign and verify an artifact with a certificate signed by trusted CA in Azure Key Vault
```

Reviewer 4: This section has a lot of overlap with the previous section. Instead of re-hashing the steps from the previous section, call out the specific differences, such as the policy change, using the csr, and creation of leaf certificate.

Author: Updated the doc structure to split getting started 2 sections into 2 files.

1. Install Azure CLI by following the [guide](https://learn.microsoft.com/en-us/cli/azure/install-azure-cli)

> 2. Login to Azure with Azure CLI, set the subscription and make sure the `GetCertificate` and `Sign` permission have been granted to your role:

Reviewer 2: ```suggestion
2. log in to Azure with Azure CLI, set the subscription, and make sure the `GetCertificate` and `Sign` permission have been granted to your role:
```

Author: Updated.

> 2. Login to Azure with Azure CLI, set the subscription and make sure the `GetCertificate` and `Sign` permission have been granted to your role:

Reviewer 2: Note: login vs log in: https://www.grammar.com/log_in_vs._login

   ```sh
   az login
   az account set --subscription $subscriptionID
   ```
3. Create an Azure Key Vault and a Certificate Signing Request (CSR):
   ```sh
   resourceGroup=notationResource
   keyVault=notationKV
   location=westus
   certName=notationLeafCert

   # create a resource group
   az group create -n $resourceGroup -l $location
   
   # create a Azure Key Vault
   az keyvault create -l $location -n $keyVault --resource-group $resourceGroup

   # generate certificate policy
   cat <<EOF > ./leafCert.json
   {
     "issuerParameters": {
       "certificateTransparency": null,
       "name": "Unknown"
     },
     "keyProperties": {
       "curve": null,
       "exportable": false,
       "keySize": 2048,
       "keyType": "RSA",
       "reuseKey": true
     },
     "secretProperties": {
       "contentType": "application/x-pem-file"
     },
     "x509CertificateProperties": {
       "ekus": [
           "1.3.6.1.5.5.7.3.3"
       ],
       "keyUsage": [
         "digitalSignature"
       ],

>        "subject": "CN=Test Signer",

Reviewer 3: similar to previous comment

Author: Updated to include 'C', 'ST', and 'O'

   # get the CSR
   CSR=$(az keyvault certificate pending show --vault-name $keyVault --name $certName --query 'csr' -o tsv)
   CSR_PATH=${certName}.csr
   printf -- "-----BEGIN CERTIFICATE REQUEST-----\n%s\n-----END CERTIFICATE REQUEST-----\n" $CSR > ${CSR_PATH}
   ```
5. please take ${certName}.csr file to a trusted CA to sign and issue your certificate, or you can use `openssl` tool to sign it locally for testing.
6. after you get the leaf certificate, you can merge the leaf certificate (`$leafCert`) to your Azure Key Vault:
   ```sh
   az keyvault certificate pending merge --vault-name $keyVault --name $certName --file $leafCert

   # get the key identifier
   keyID=$(az keyvault certificate show -n $certName --vault-name $keyVault --query 'kid' -o tsv)
   ```
7. run a local registry and push an image to be signed:
   ```sh
   # run a local registry with Docker
   docker run --rm -d -p 5000:5000 ghcr.io/oras-project/registry:v1.0.0-rc.4
   
   # push a hello-world image
   docker pull hello-world
   docker tag hello-world:latest localhost:5000/hello-world:v1
   docker push localhost:5000/hello-world:v1
   ```

> 8. notation sign with an external certificate bundle (`$certBundlePath`) including the intermediate certificates and a root certificate in PEM format. You may fetch the certificate bundle from your CA official website.

Author: Updated.

> 6. Prepare a container registry. You can follow the [guide](https://learn.microsoft.com/en-us/azure/container-registry/container-registry-get-started-portal?tabs=azure-cli) to create an Azure Container Registry which is recommended. Suppose the logging server is `notation.azurecr.io` with an `hello-world:v1` image.

Author: It's a typo. I want to say "login server" which is the host name of the registry.

>    notation sign notation.azurecr.io/hello-world:v1 --id $keyID --plugin azure-kv --plugin-config=ca_certs=$certBundlePath

Reviewer 3: need to login to the registry first

Author: Updated.

   ```sh
   # sign with azure-kv
   notation sign localhost:5000/hello-world:v1 --id $keyID --plugin azure-kv --plugin-config=ca_certs=$certBundlePath
   # example output
   Warning: Always sign the artifact using digest(@sha256:...) rather than a tag(:v1) because tags are mutable and a tag reference can point to a different artifact than the one signed.
   Successfully signed localhost:5000/hello-world@sha256:f54a58bc1aac5ea1a25d796ae155dc228b3f0e11d046ae276b39c4bf2f13d8c4
   ```

> 9. notation verify command needs the root certificate of your CA in the trust store:

Reviewer 2: ```suggestion
9. Signature verification with Notation needs the root certificate of your CA in the trust store:
```

Author: Updated.

Azure Provider for the [Notation CLI](https://github.com/notaryproject/notation)

The `notation-azure-kv` plugin provides the capability to sign the Notation generated payload by using Azure Key Vault (AKV). The user's certificate and private key should be stored in AKV and the plugin will request signing and getting the leaf certificate from AKV. 

The plugin supports Azure CLI identity and Managed Identity for accessing AKV.

## Installation
Before installing notation azure key vault plugin, please make sure Notation CLI has been installed. If you didn't install it, please follow the [Notation installation guide](https://notaryproject.dev/docs/installation/cli/).

1. Navigate to the [Releases](https://github.com/Azure/notation-azure-kv/releases) page and select the latest release of `notation-azure-kv`. The `notation-azure-kv` binaries for each platform are available under the Assets section. Please download it based on your own platform. 
2. Validate the checksum which should be the same as the one in `checksums.txt` and then install the plugin.
   
   For Linux Bash:
   ```sh
   version=0.6.0

   # validate checksum

>    sha256sum notation-azure-kv_${version}_linux_amd64.tar.gz

Reviewer: Comparing the hash value with `checksum.txt` should not be done by human eyes.

Think about
```sh
cat checksums.txt | grep notation-azure-kv_${version}_linux_amd64.tar.gz | sha256sum -c
```

Author: Updated.

   # install the plugin
   mkdir -p "$HOME/.config/notation/plugins/azure-kv"
   tar zxf notation-azure-kv_<version>_linux_amd64.tar.gz -C "$HOME/.config/notation/plugins/azure-kv" notation-azure-kv
   ```

>    For MacOS Zsh:

Author: Updated.

[Microsoft's Trademark & Brand Guidelines](https://www.microsoft.com/en-us/legal/intellectualproperty/trademarks/usage/general).
Use of Microsoft trademarks or logos in modified versions of this project must not cause confusion or imply Microsoft sponsorship.
Any use of third-party trademarks or logos are subject to those third-party's policies.

> ~~~~

Reviewer: This line seems redundant.

Author: Removed.
